compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -856,12 +856,25 @@ impl<'tcx> Cx<'tcx> {
 
             Res::Def(DefKind::ConstParam, def_id) => {
                 let hir_id = self.tcx.hir().local_def_id_to_hir_id(def_id.expect_local());
-                let generics = self.tcx.generics_of(hir_id.owner);
-                let index = generics.param_def_id_to_index[&def_id];
-                let name = self.tcx.hir().name(hir_id);
-                let param = ty::ParamConst::new(index, name);
-
-                ExprKind::ConstParam { param, def_id }
+                if let Some(index) =
+                    self.tcx.generics_of(hir_id.owner).param_def_id_to_index.get(&def_id)
+                {
+                    let name = self.tcx.hir().name(hir_id);
+                    let param = ty::ParamConst::new(*index, name);
+                    ExprKind::ConstParam { param, def_id }
+                } else {
+                    use rustc_middle::middle::resolve_bound_vars as rbv;
+                    match self.tcx.named_bound_var(expr.hir_id) {
+                        Some(rbv::ResolvedArg::LateBound(_, index, _)) => {
+                            let name = self.tcx.item_name(def_id);
+                            let param = ty::ParamConst::new(index, name);
+                            ExprKind::ConstParam { param, def_id }
+                        }
+                        arg => {
+                            bug!("unexpected bound var resolution for {:?}: {arg:?}", expr.hir_id)
+                        }
+                    }
+                }
```

Comment:
This code should be this one
https://github.com/rust-lang/rust/blob/a9842c73bcd78ca4c2ec56f3e529aed79bd37df7/compiler/rustc_hir_analysis/src/astconv/mod.rs#L2675-L2693

but with `Const` instead of `Ty` where relevant.

Reply:
Constructing a `ParamConst` like the code is currently doing is incorrect, and is problematic because we'll be treating it like an early-bound const param.

There currently isn't a THIR representation for a bound const that's been "liberated" like a placeholder is. I'm not really sure what that would look like, either. This probably needs a more involved solution than what's curently implemented.